src/lib/storage/dictionary_column.hpp

```diff
@@ -0,0 +1,144 @@
+  ValueID lower_bound(T value) const { return _to_value_id(value); }
+
+  // same as lower_bound(T), but accepts an AllTypeVariant
+  ValueID lower_bound(const AllTypeVariant& value) const { return _to_value_id(type_cast<T>(value)); }
+
+  // returns the first value ID that refers to a value > the search value
+  // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
+  ValueID upper_bound(T value) const {
+    const auto found = std::upper_bound(_dictionary->cbegin(), _dictionary->cend(), value);
+    if (found == _dictionary->cend()) {
+      return INVALID_VALUE_ID;
+    }
+    return ValueID(found - _dictionary->cbegin());
+  }
+
+  // same as upper_bound(T), but accepts an AllTypeVariant
+  ValueID upper_bound(const AllTypeVariant& value) const { return upper_bound(type_cast<T>(value)); }
+
+  // return the number of unique_values (dictionary entries)
+  size_t unique_values_count() const { return _dictionary->size(); }
+
+  // return the number of entries
+  size_t size() const override { return _attribute_vector->size(); }
+
+ protected:
+  std::shared_ptr<std::vector<T>> _dictionary;
+  std::shared_ptr<BaseAttributeVector> _attribute_vector;
+
+ protected:
+  void _build_dictionary(const std::shared_ptr<const BaseColumn>& base_column) {
+    _copy_values(base_column);
+    std::sort(_dictionary->begin(), _dictionary->end());
+    _dictionary->erase(std::unique(_dictionary->begin(), _dictionary->end()), _dictionary->end());
+    _dictionary->shrink_to_fit();
+  }
+
+  void _copy_values(const std::shared_ptr<const BaseColumn>& base_column) {
+    const auto value_column = std::dynamic_pointer_cast<const ValueColumn<T>>(base_column);
+    if (value_column) {
+      _copy_values_fast(value_column);
+    } else {
+      _copy_values_slow(base_column);
+    }
+  }
+
+  void _copy_values_fast(const std::shared_ptr<const ValueColumn<T>>& value_column) {
+    (*_dictionary) = value_column->values();
+  }
+
+  void _copy_values_slow(const std::shared_ptr<const BaseColumn>& base_column) {
+    PerformanceWarning("element-wise copy of column");
+    for (size_t index = 0; index < base_column->size(); ++index) {
+      _dictionary->emplace_back(type_cast<T>((*base_column)[index]));
+    }
+  }
+
+  void _assign_attribute_vector(const size_t size) {
+    Assert(size < std::numeric_limits<uint64_t>::max(), "Number of attributes out of range");
+    if (size < std::numeric_limits<uint8_t>::max()) {
+      _attribute_vector = std::make_shared<FittedAttributeVector<uint8_t>>(size);
+    } else if (size < std::numeric_limits<uint16_t>::max()) {
+      _attribute_vector = std::make_shared<FittedAttributeVector<uint16_t>>(size);
+    } else if (size < std::numeric_limits<uint32_t>::max()) {
+      _attribute_vector = std::make_shared<FittedAttributeVector<uint32_t>>(size);
+    } else {
+      _attribute_vector = std::make_shared<FittedAttributeVector<uint64_t>>(size);
```

Comment:
`uint64_t` should only be used for the first part of the implementation, in the end the allowed template types were `uint8_t`, `uint16_t`, `uint32_t`. `uint32_t` will probably be enough to adress all elements (at least in our academic use case)

Reply:
OK.